dot/rpc/modules/state.go

```diff
@@ -403,30 +404,71 @@ func (sm *StateModule) GetStorageSize(
 	return nil
 }
 
-// QueryStorage isn't implemented properly yet.
+// QueryStorage Query historical storage entries (by key) starting from a request start block given,
+//until request end block or best block if end block in nil
 func (sm *StateModule) QueryStorage(
 	_ *http.Request, req *StateStorageQueryRangeRequest, res *[]StorageChangeSetResponse) error {
 	if req.StartBlock.IsEmpty() {
-		return errors.New("the start block hash cannot be an empty value")
+		return ErrStartBlockValueEmpty
 	}
 
-	changesByBlock, err := sm.coreAPI.QueryStorage(req.StartBlock, req.EndBlock, req.Keys...)
+	startBlock, err := sm.blockAPI.GetBlockByHash(req.StartBlock)
 	if err != nil {
 		return err
 	}
 
-	response := make([]StorageChangeSetResponse, 0, len(changesByBlock))
+	startBlockNumber := startBlock.Header.Number
 
-	for block, c := range changesByBlock {
-		var changes [][]string
+	var endBlockHash common.Hash
+	if req.EndBlock.IsEmpty() {
+		endBlockHash = sm.blockAPI.BestBlockHash()
+	} else {
+		endBlockHash = req.EndBlock
+	}
+	endBlock, err := sm.blockAPI.GetBlockByHash(endBlockHash)
+	if err != nil {
+		return err
+	}
+	endBlockNumber := endBlock.Header.Number
 
-		for key, value := range c {
-			changes = append(changes, []string{key, value})
+	response := make([]StorageChangeSetResponse, 0)
+	lastValue := make([]*string, len(req.Keys))
+	firstPass := true
+	for i := startBlockNumber; i < endBlockNumber; i++ {
+		bHash, err := sm.blockAPI.GetHashByNumber(i)
+		if err != nil {
+			return err
 		}
+		var changes [][]*string
 
-		blochHash := block
+		for j, key := range req.Keys {
+			value, err := sm.storageAPI.GetStorageByBlockHash(&bHash, common.MustHexToBytes(key))
```

Comment:
I've named this GetStorage because the RPC methods are named GetStorage, so this name was used to keep consistent with the names defined in the PSP-6 spec for RPC methods.

Reply:
That's a quite specific and very valid answer!! 👍